src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/VisitService.kt

```diff
@@ -33,8 +36,100 @@ class VisitService(
+        }
+        visitEntity.support.add(createVisitSupport(visitEntity, it.type, it.text))
+      }
+    }
+
+    createVisitTrackEvent(visitEntity)
+
+    return VisitDto(visitEntity)
+  }
+
+  fun changeReservedVisitSlot(applicationReference: String, changeReservedVisitSlotRequestDto: ChangeReservedVisitSlotRequestDto): VisitDto {
+    val visitEntity = visitRepository.findByApplicationReference(applicationReference) ?: throw VisitNotFoundException("Reserved visit reference $applicationReference not found")
+
+    changeReservedVisitSlotRequestDto.visitRestriction?.let { visitRestriction -> visitEntity.visitRestriction = visitRestriction }
+    changeReservedVisitSlotRequestDto.startTimestamp?.let { visitStart -> visitEntity.visitStart = visitStart }
+    changeReservedVisitSlotRequestDto.endTimestamp?.let { visitEnd -> visitEntity.visitEnd = visitEnd }
+
+    changeReservedVisitSlotRequestDto.visitContact?.let { visitContactUpdate ->
+      visitEntity.visitContact?.let { visitContact ->
+        visitContact.name = visitContactUpdate.name
+        visitContact.telephone = visitContactUpdate.telephone
+      } ?: run {
+        visitEntity.visitContact = createVisitContact(visitEntity, visitContactUpdate.name, visitContactUpdate.telephone)
+      }
+    }
+
+    changeReservedVisitSlotRequestDto.visitors?.let { visitorsUpdate ->
+      visitEntity.visitors.clear()
+      visitRepository.saveAndFlush(visitEntity)
+      visitorsUpdate.distinctBy { it.nomisPersonId }.forEach {
+        visitEntity.visitors.add(createVisitVisitor(visitEntity, it.nomisPersonId, it.visitContact))
+      }
+    }
+
+    changeReservedVisitSlotRequestDto.visitorSupport?.let { visitSupportUpdate ->
+      visitEntity.support.clear()
+      visitRepository.saveAndFlush(visitEntity)
+      visitSupportUpdate.forEach {
+        if (!supportTypeRepository.existsByName(it.type)) {
+          throw SupportNotFoundException("Invalid support ${it.type} not found")
+        }
+        visitEntity.support.add(createVisitSupport(visitEntity, it.type, it.text))
+      }
+    }
+
+    telemetryClient.trackEvent(
+      "visit-scheduler-prison-visit-updated",
+      listOfNotNull(
+        "reference" to visitEntity.reference,
+        "prisonerId" to visitEntity.prisonerId,
+        "prisonId" to visitEntity.prisonId,
+        "visitType" to visitEntity.visitType.name,
+        "visitRoom" to visitEntity.visitRoom,
+        if (changeReservedVisitSlotRequestDto.visitRestriction != null) "visitRestriction" to changeReservedVisitSlotRequestDto.visitRestriction.name else null,
+        if (changeReservedVisitSlotRequestDto.startTimestamp != null) "visitStart" to changeReservedVisitSlotRequestDto.startTimestamp.format(DateTimeFormatter.ISO_DATE_TIME) else null,
+        if (visitEntity.visitStatus != null) "visitStatus" to visitEntity.visitStatus.name else null
+      ).toMap(),
+      null
+    )
+
+    return VisitDto(visitEntity)
+  }
+
+  @Suppress("KotlinDeprecation")
```

Comment:
Should we add @Deprecated annotation here if being called only from LegacyController?

Reply:
@Suppress("KotlinDeprecation") is the same thing!